### PR TITLE
My nc news 7

### PR DESCRIPTION
**Pull request by kamiviolet**
1. update endpoints.json
2. add new method to endpoint /api/articles/:article_id/comments
3. add test suite
- 2 tests for status 201
- 5 tests for status 400 (as it is a POST method, i decided to use 400 even if the numeric article_id is not found as it is not to GET the article but a REQUEST)

updated..

### CODE REVIEW

File: __tests__/app.test.js [hunk 1]
```diff
@@ -218,6 +221,7 @@ describe('/api/articles/:article_id/comments', () => {
                 expect(message).toBe('Bad request.')
         })
     })
+
     it('GET - status 404 - invalid numeric article_id will respond with not found.', () => {
```
Comment (yalicee) [inline __tests__/app.test.js]: technically this is a valid numeric article but just non existent

File: __tests__/app.test.js [hunk 2]
```diff
@@ -227,4 +231,105 @@ describe('/api/articles/:article_id/comments', () => {
...
+                    .query(`
+                        SELECT * FROM comments
+                        WHERE author = $1 AND body = $2 AND article_id = $3;
+                    `, [example.username, example.body, article_id])
+            })
+            .then(({rows}) => {
+                expect(rows.length).toBe(1);
+                expect(rows[0].author).toBe(example.username);
+                expect(rows[0].body).toBe(example.body);
+                expect(rows[0].article_id).toBe(article_id);
```
Comment (yalicee) [inline __tests__/app.test.js]: SQL is not needed in testing. This is similar to the test you have above as you return back the comment object anyway. Please remove

File: __tests__/app.test.js [hunk 3]
```diff
@@ -227,4 +231,105 @@ describe('/api/articles/:article_id/comments', () => {
...
+    it('POST - status 400 - no request body is sent.', () => {
+        return request(app)
+            .post('/api/articles/2/comments')
+            .send({})
+            .expect(400)
+            .then(({body}) => {
+                const {message} = body;
+                expect(message).toBe('It seems you forget to send the request...')
+        })
+    })
```
Comment (yalicee) [inline __tests__/app.test.js]: This test could be considered redundant due to test above

File: __tests__/app.test.js [hunk 4]
```diff
@@ -227,4 +231,105 @@ describe('/api/articles/:article_id/comments', () => {
...
+    it('POST - status 400 - the article_id is not existing.', () => {
+        return request(app)
+            .post('/api/articles/99999/comments')
+            .send({username: 'XXX', body: 'XXXX'})
+            .expect(400)
+            .then(({body}) => {
+                const {message} = body;
+                expect(message).toBe('Apology. Non-existing article_id.')
+        })
+    })
```
Comment (yalicee) [inline __tests__/app.test.js]: This is a 404 not found test

File: __tests__/app.test.js [hunk 5]
```diff
@@ -227,4 +231,105 @@ describe('/api/articles/:article_id/comments', () => {
...
+
+    it('POST - status 400 - the username is not existing.', () => {
+        return request(app)
+            .post('/api/articles/2/comments')
+            .send({username: 'XXX', body: 'XXXX'})
+            .expect(400)
+            .then(({body}) => {
+                const {message} = body;
+                expect(message).toBe('Apology. Non-existing username, please sign up to leave your comment.')
+        })
```
Comment (yalicee) [inline __tests__/app.test.js]: Same for this one

File: controllers/comments.controller.js [hunk 6]
```diff
@@ -12,4 +12,25 @@ const getCommentsByArticleId = (req, res, next) => {
...
+    const {article_id} = req.params;
+    if (Object.keys(newComment).length === 0) {
+        const err = {status: 400, message: 'It seems you forget to send the request...'}
+        next(err);
+    } else if (!Object.hasOwn(newComment, 'username') || !Object.hasOwn(newComment, 'body')) {
+        const err = {status: 400, message: 'Comment not in correct format.'}
+        next(err);
+    } else if (Number.isNaN(+article_id) !== false) {
+        const err = {status: 400, message: 'Invalid article_id.'}
+        next(err);
```
Comment (yalicee) [inline controllers/comments.controller.js]: Have a think about if you can move this functionality somewhere else? Perhaps in your error handlers?

File: db/seeds/utils.js [hunk 7]
```diff
@@ -32,7 +32,21 @@ exports.validateExistingId = (id) => {
...
+exports.validateExistingUser = (username) => {
+  return db
+    .query(`
+      SELECT * 
+      FROM users
+      WHERE username in ($1)
+    `, [username])
+    .then(({rows}) => {
+      if (rows.length === 0) {
+        return {status: 404, message: 'The username is currently not found.'}
```
Comment (yalicee) [inline db/seeds/utils.js]: author in the comments table is a foreign key which relates to username in the users table. A psql error will be thrown if this is violated. It is not necessary to separately check if the user exists. same for article_id

File: models/comments.model.js [hunk 8]
```diff
@@ -1,16 +1,45 @@
...
+}
+
+exports.createNewCommentByArticleId = (id, comment) => {
+    return validateExistingId(id)
+        .then((msg) => {
+            if (msg !== undefined) {
+                return Promise.reject({status: 400, message: 'Apology. Non-existing article_id.'})
+            }
+        })
+        .then(() => validateExistingUser(comment.username))
```
Comment (yalicee) [inline models/comments.model.js]: Again no need to check if they exist

**Review by yalicee — CHANGES_REQUESTED** [review]
Other things to consider besides my comments:
-article_id is also a foreign key so again, a separate check that the article exists should not be necessary as a psql error will be thrown.
-a psql error will also be thrown if article_id is not a number.

**Review by yalicee — CHANGES_REQUESTED** [review]
Other things to consider besides my comments:
-article_id is also a foreign key so again, a separate check that the article exists should not be necessary as a psql error will be thrown.
-a psql error will also be thrown if article_id is not a number.

**Comment by kamiviolet** [comment]
> Other things to consider besides my comments: -article_id is also a foreign key so again, a separate check that the article exists should not be necessary as a psql error will be thrown. -a psql error will also be thrown if article_id is not a number.

@yalicee Sorry i am stuck... i added the extra util function to check if its a valid foreign key because if not doing so, my terminal is hanging from Jest 500s timeout... how should i organise to be able to achieve so? T_T

I tried to add extra catch appended to the db.query in model file.. (not recommended to have catch everywhere but thats the only way to avoid hanging and timeout) but then the controller can log the psql error but not able to get to my error handler in app.js... how so? hm...

Otherwise I have adjusted the status and statements to be more accurately reflecting to the situation accordingly

File: __tests__/app.test.js [hunk 9]
```diff
@@ -218,13 +221,82 @@ describe('/api/articles/:article_id/comments', () => {
...
+            .then(({body}) => {
+                const {message} = body;
+                expect(message).toBe('Invalid request format.')
+        })
+    })
+
+    it('POST - status 404 - the article_id is not existing.', () => {
+        return request(app)
+            .post('/api/articles/99999/comments')
+            .send({username: 'XXX', body: 'XXXX'})
```
Comment (yalicee) [inline __tests__/app.test.js]: Please change this to be a valid comment object so it's less confusing what error it is for

File: app.js [hunk 10]
```diff
@@ -21,20 +21,32 @@ app.route('/api/articles/:article_id')
...
+    if (err.code === '23503') {
+        res.status(404).send({message: 'Request value does not exist at the moment in database.'});
+    }
+    
+    if (err.status && err.message) {
         res.status(err.status).send({message: err.message});
     }
+
+    res.status(400).send({message: 'Bad request.'});
 })
```
Comment (yalicee) [inline app.js]: There is an opportunity to refactor this later on

File: controllers/comments.controller.js [hunk 11]
```diff
@@ -12,4 +12,17 @@ const getCommentsByArticleId = (req, res, next) => {
...
 
-module.exports = { getCommentsByArticleId }
\ No newline at end of file
+const postNewCommentByArticleId = (req, res, next) => {
+    const newComment = req.body;
+    const {article_id} = req.params;
+
+    return createNewCommentByArticleId(article_id, newComment)
+        .then((comment) => {
+            res.status(201).send({comment: comment[0]})
```
Comment (yalicee) [inline controllers/comments.controller.js:21]: would prefer you to index this array in the model instead of the controller

**Review by yalicee — APPROVED** [review]

**Pull request merged**